bigquery/storage/managedwriter/adapt/protoconversion.go

```diff
@@ -0,0 +1,260 @@
+	"google.golang.org/protobuf/types/descriptorpb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+// bqModeToFieldLabelMap holds mapping from field schema mode to proto label.
+// proto3 no longer allows use of REQUIRED labels, so we solve that elsewhere
+// and simply use optional.
+var bqModeToFieldLabelMap = map[storagepb.TableFieldSchema_Mode]descriptorpb.FieldDescriptorProto_Label{
+	storagepb.TableFieldSchema_NULLABLE: descriptorpb.FieldDescriptorProto_LABEL_OPTIONAL,
+	storagepb.TableFieldSchema_REPEATED: descriptorpb.FieldDescriptorProto_LABEL_REPEATED,
+	storagepb.TableFieldSchema_REQUIRED: descriptorpb.FieldDescriptorProto_LABEL_OPTIONAL,
+}
+
+// Allows conversion between BQ schema type and FieldDescriptorProto's type.
+var bqTypeToFieldTypeMap = map[storagepb.TableFieldSchema_Type]descriptorpb.FieldDescriptorProto_Type{
+	storagepb.TableFieldSchema_BIGNUMERIC: descriptorpb.FieldDescriptorProto_TYPE_BYTES,
+	storagepb.TableFieldSchema_BOOL:       descriptorpb.FieldDescriptorProto_TYPE_BOOL,
+	storagepb.TableFieldSchema_BYTES:      descriptorpb.FieldDescriptorProto_TYPE_BYTES,
+	storagepb.TableFieldSchema_DATE:       descriptorpb.FieldDescriptorProto_TYPE_INT32,
+	storagepb.TableFieldSchema_DATETIME:   descriptorpb.FieldDescriptorProto_TYPE_INT64,
+	storagepb.TableFieldSchema_DOUBLE:     descriptorpb.FieldDescriptorProto_TYPE_DOUBLE,
+	storagepb.TableFieldSchema_GEOGRAPHY:  descriptorpb.FieldDescriptorProto_TYPE_STRING,
+	storagepb.TableFieldSchema_INT64:      descriptorpb.FieldDescriptorProto_TYPE_INT64,
+	storagepb.TableFieldSchema_NUMERIC:    descriptorpb.FieldDescriptorProto_TYPE_BYTES,
+	storagepb.TableFieldSchema_STRING:     descriptorpb.FieldDescriptorProto_TYPE_STRING,
+	storagepb.TableFieldSchema_STRUCT:     descriptorpb.FieldDescriptorProto_TYPE_MESSAGE,
+	storagepb.TableFieldSchema_TIME:       descriptorpb.FieldDescriptorProto_TYPE_INT64,
+	storagepb.TableFieldSchema_TIMESTAMP:  descriptorpb.FieldDescriptorProto_TYPE_INT64,
+}
+
+// For TableFieldSchema OPTIONAL mode, we use the wrapper types to allow for the
+// proper representation of NULL values, as proto3 semantics would just use default value.
+var bqTypeToWrapperMap = map[storagepb.TableFieldSchema_Type]string{
+	storagepb.TableFieldSchema_BIGNUMERIC: ".google.protobuf.BytesValue",
+	storagepb.TableFieldSchema_BOOL:       ".google.protobuf.BoolValue",
+	storagepb.TableFieldSchema_BYTES:      ".google.protobuf.BytesValue",
+	storagepb.TableFieldSchema_DATE:       ".google.protobuf.Int32Value",
+	storagepb.TableFieldSchema_DATETIME:   ".google.protobuf.Int64Value",
+	storagepb.TableFieldSchema_DOUBLE:     ".google.protobuf.DoubleValue",
+	storagepb.TableFieldSchema_GEOGRAPHY:  ".google.protobuf.StringValue",
+	storagepb.TableFieldSchema_INT64:      ".google.protobuf.StringValue",
+	storagepb.TableFieldSchema_NUMERIC:    ".google.protobuf.BytesValue",
+	storagepb.TableFieldSchema_STRING:     ".google.protobuf.StringValue",
```

Comment:
Do we need this wrapper?

Reply:
empty string vs null string?

Reply:
String field can be not set, which means null?